[WFCORE-4291] Add support for non-graceful startup

jasondlee opened the pull request:
Add test for non-graceful server start
Add test applications

Modify Host to handle state where queueRequestsOnStart is not explicitly configured.
Take into account "non-graceful" status when deciding whether to queue request or not

Do not use GateHandlerWrapper if we're running in non-graceful/non-enqueued mode

Minor code cleanup

https://issues.redhat.com/browse/WFCORE-4291

DarkTatka commented on undertow/src/main/java/org/wildfly/extension/undertow/Host.java:
```diff
@@ -133,26 +135,42 @@ private String getDeployedContextPath(DeploymentInfo deploymentInfo) {
+                UndertowLogger.ROOT_LOGGER.debug("Running in non-graceful mode and " + QUEUE_REQUESTS_ON_START.getName() +
+                        " not explicitly set. Requests will not be queued.");
+            } else {
+                if (queueRequestsOnStart == null || Boolean.TRUE == queueRequestsOnStart) {
```
I suggest using `Boolean.TRUE.equals(queueRequestsOnStart)` to compare values, not instance references.

jasondlee replied:
Bah. Good catch. :)

bstansberry commented on docs/src/main/asciidoc/_admin-guide/management-tasks/Suspend,_Resume_and_Graceful_shutdown.adoc:
```diff
@@ -232,3 +232,42 @@ the proper way to control the lifecycle of a domain mode server process is via t
+[[non-graceful-startup]]
+== Non-graceful Startup
+
+By default, Wildfly starts up gracefully, meaning that incoming requests are queued until the
```
s/queued/queued or cleanly rejected/g

The queuing behavior is limited to HTTP requests hitting the undertow listener, and only if queuing is enabled on that listener.

bstansberry replied:
Also

s/Wildfly/WildFly

bstansberry commented on docs/src/main/asciidoc/_admin-guide/management-tasks/Suspend,_Resume_and_Graceful_shutdown.adoc:
```diff
@@ -232,3 +232,42 @@ the proper way to control the lifecycle of a domain mode server process is via t
+start, as Application A is waiting for data from Application B before its deploy/startup can
+complete. In this situation, a deadlock occurs, and the server startup times out.
+
+A non-graceful startup is intended to address this situation in that it will allow Wildfly to
```
s/Wildfly/WildFly

bstansberry commented on docs/src/main/asciidoc/_admin-guide/management-tasks/Suspend,_Resume_and_Graceful_shutdown.adoc:
```diff
@@ -232,3 +232,42 @@ the proper way to control the lifecycle of a domain mode server process is via t
+Application B has successfully deployed/started, Application A can also start immediately, as its
+request will be fulfilled. Note, however, that a race condition can occur: if Application B is
+not yet deployed (e.g., the deploy order is incorrect, or B has not finished starting), then
+Application A may still fail to start since Application B is not available. Wildfly users making
```
s/Wildfly/WildFly

bstansberry commented on docs/src/main/asciidoc/_admin-guide/management-tasks/Suspend,_Resume_and_Graceful_shutdown.adoc:
```diff
@@ -232,3 +232,42 @@ the proper way to control the lifecycle of a domain mode server process is via t
+not yet deployed (e.g., the deploy order is incorrect, or B has not finished starting), then
+Application A may still fail to start since Application B is not available. Wildfly users making
+use of non-graceful startups must be aware of this and take steps to remedy those scenarios. With
+a non-graceful startup, however, Wildfly will no longer be the cause of a deployment failure in
```
s/Wildfly/WildFly

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Undertow.adoc:
```diff
@@ -247,6 +247,15 @@ be matched
+Note: If a <<non-graceful-startup>> is requested, and the queue-requests-on-start attribute is not set, requests will NOT be
+queued despite the default value of true for the property. In the instance of a non-graceful startup, non-queued requests
+are required. However, if non-graceful is configured, but queue-requests-on-start is explicitly set to true, then requests
+will be queued, effectively disabling the non-graceful mode
```
"effectively disabling the non-graceful mode for requests to that host."

bstansberry commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/server/nongraceful/deploymenta/ResourceA.java:
```diff
@@ -0,0 +1,15 @@
+public class ResourceA {
+    @GET
+    public String get() {
+        System.out.println("***** [non-graceful] Inside ResourceA");
```
Please comment out all of these. We want to avoid logging during tests as it puts IO pressure on the CI servers.

jasondlee replied:
I forgot all about those. :)

bstansberry commented:
@jasondlee This needs to have the top commit from https://github.com/bstansberry/wildfly/commits/start_when_running added in.

jasondlee commented:
> @jasondlee This needs to have the top commit from https://github.com/bstansberry/wildfly/commits/start_when_running added in.

I have cherry-picked and pushed the changes.

jasondlee committed:
[WFCORE-4291] Add support for non-graceful startup

Update parent pom to pick up current core changes
Modify Host to handle state where queueRequestsOnStart is not explicitly configured.
Take into account "non-graceful" status when deciding whether to queue request or not
Do not use GateHandlerWrapper if we're running in non-graceful/non-enqueued mode
Cherry-pick Brian's commit: fe5d72510159c3a968f36f2151344b9bd72d0e66
Add missing module dependencies

Add test for non-graceful server start
Add test applications

Add docs for new command line option
Add section describing non-graceful startup

Update community docs
Modify Undertow documentation to describe interaction with --graceful-startup

bstansberry commented on batch-jberet/src/main/java/org/wildfly/extension/batch/jberet/deployment/JobOperatorService.java:
```diff
@@ -432,7 +448,27 @@ public void suspended(final ServerActivityCallback serverActivityCallback) {
+            if (doResume) {
+                restartStoppedJobs();
+            }
         }
```
@chengfang FYI. What this is about is the fact that with this RFE the behavior of boot can be different, where users can configure 'graceful startup' to be turned off. With graceful startup this kind of sequence happens

server is initially in a suspended state
boot is 99% done, boot logic 'resumes' the server to move out of suspended. The resume() method above is called
server switches from ControlledProcessState.State.STARTING to RUNNING (i.e. boot is done)

The gap between the last two bits is miniscule.

If the user has disable graceful startup, the server is not initially in suspended state, so no resume call goes out. So no signal that triggers 'restartStoppedJobs' will be received. So this change is adding a listener for when the ControlledProcessState.State.STARTING to RUNNING change happens and uses that to ensure  restartStoppedJobs is called.

If after boot the server is suspend and later resumed, that will result in a call to restartStoppedJobs, same as before.

bstansberry commented on transactions/src/main/java/org/jboss/as/txn/suspend/RecoverySuspendController.java:
```diff
@@ -48,9 +64,45 @@ public void suspended(ServerActivityCallback serverActivityCallback) {
+            resumeRecovery();
+        }
+
+    }
```
@mmusgrov @ochaloup  FYI. What this is about is the fact that with this RFE the behavior of boot can be different, where users can configure 'graceful startup' to be turned off. With graceful startup this kind of sequence happens

server is initially in a suspended state as boot proceeds
boot is 99% done, boot logic 'resumes' the server to move out of suspended. The resume() method above is called
server switches from ControlledProcessState.State.STARTING to RUNNING (i.e. boot is done)

The gap between the last two bits is miniscule.

If the user has disable graceful startup, the server is not initially in suspended state, so no resume call goes out. So no signal that triggers 'recoveryManagerService.resume()' will be received. So this change is adding a listener for when the ControlledProcessState.State.STARTING to RUNNING change happens and uses that to ensure recoveryManagerService.resume() is called.

If after boot the server is suspend and later resumed, that will result in a call to recoveryManagerService.resume(), same as before.

ochaloup reviewed (approved):
This seems fine to me.

honza-kasik commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/server/nongraceful/NongracefulStartTestCase.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @Test
+    public void testNonGracefulDeployment() {
```
Can you please describe the test briefly and passing criteria? It could IMHO help other maintainers.

jasondlee replied:
Done.

honza-kasik commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/server/nongraceful/NongracefulStartTestCase.java:
```diff
@@ -0,0 +1,78 @@
+            logger.trace("Exception during container cleanup and shutdown", e);
+        }
+    }
+}
```
Maybe a newline missing?

honza-kasik commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/server/nongraceful/deploymenta/ResourceA.java:
```diff
@@ -0,0 +1,14 @@
+    public String get() {
+        return "Hello";
+    }
+}
```
Typo: newline missing.

honza-kasik commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/server/nongraceful/deploymenta/TestApplicationA.java:
```diff
@@ -0,0 +1,13 @@
+@ApplicationPath("/testa")
+public class TestApplicationA extends Application {
+
+}
```
Typo: newline missing.

honza-kasik commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/server/nongraceful/deploymentb/ResourceB.java:
```diff
@@ -0,0 +1,42 @@
+    public String get() {
+        return "Hello";
+    }
+}
```
Typo: newline missing.

honza-kasik commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/server/nongraceful/deploymentb/TestApplicationB.java:
```diff
@@ -0,0 +1,12 @@
+@ApplicationPath("/testb")
+public class TestApplicationB extends Application {
+
+}
```
Typo: newline missing.

honza-kasik reviewed (commented):
Thank you @jasondlee ! I reviewed _only the documentation and tests_ and left few very minor comments behind.

jasondlee commented:
/retest
